File: Interface/source/State.cpp
#include "State.h"
#include "IO.h"

State::State(const UMeshDataNative udata) {
	VSize = udata.VSize;
	FSize = udata.FSize;

	V = new Eigen::MatrixXf(VSize, 3);
	N = new Eigen::MatrixXf(VSize, 3);
	C = new Eigen::MatrixXf(VSize, 4);
	UV = new Eigen::MatrixXf(VSize, 2);
	F = new Eigen::MatrixXi(FSize, 3);

	S = new Eigen::VectorXi(VSize);

	// Copy over data
	TransposeFromMap(udata.VPtr, V);
	TransposeFromMap(udata.NPtr, N);
	TransposeFromMap(udata.CPtr, C);
	TransposeFromMap(udata.UVPtr, UV);
	TransposeFromMap(udata.FPtr, F);

	// Selection
	S->setZero();
	DirtySelections = (unsigned int)-1;

	SSize = new unsigned int[32];
	std::fill(SSize, SSize + 32, 0);

	Native = new NativeState{};
}

State::~State() {
	delete V;
	delete N;
	delete C;
	delete UV;
	delete F;
	delete S;
	delete[] SSize;
	delete Native;
}

File: Assets/Scripts/Libigl/State.cs
using System.Runtime.InteropServices;

namespace Libigl
{
    /// <summary>
    /// The C++ state for a mesh, Stores only pointers so this can be shared between C++ and C#.<br/>
    /// A subset of the variables represents the ColMajor version of <see cref="UMeshData"/> and the <see cref="DirtyState"/>
    /// indicates which of these have been changed and need to be applied to the Unity mesh.<br/>
    /// As this struct is shared between a managed (C#) and native (C++) context you must consider Marshalling when adding new variables.
    /// <remarks><c>void*</c> usually represents an Eigen matrix.</remarks>
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public unsafe struct State
    {
        // ColMajor Mesh data Shared with UnityMeshData
        public uint DirtyState;
        public uint DirtySelections;
        public uint DirtySelectionsResized;

        public readonly void* VPtr;
        public readonly void* NPtr;
        public readonly void* CPtr;
        public readonly void* UVPtr;
        public readonly void* FPtr;

        public readonly int VSize;
        public readonly int FSize;

        public readonly void* S;
        public readonly uint* SSize; // uint[32], vertices selected per selection
        public readonly uint SSizeAll; // Total vertices selected
        public uint SCount; // Amount of selections

        // Native only state
        private readonly void* Native;
    }
}